### ISSUE TITLE
GPU - unable minimize loss / CPU - behaves normal

### ISSUE THREAD

**Issue by codyfalkosky** [tensorflow/tensorflow#61772]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.13.0

### Custom code

Yes

### OS platform and distribution

OSX 12.4

### Python version

3.11.4

### Current behavior?

With all other factors the same, GPU does not minimize loss and CPU does.
![GPU_behavoir](https://github.com/tensorflow/tensorflow/assets/130612303/33ff3317-94e2-48f0-9f5c-dadedc48e54c)
![CPU_behavior](https://github.com/tensorflow/tensorflow/assets/130612303/28b04dc4-2271-497f-a285-80b87835a29e)

### Standalone code to reproduce the issue

```shell
link to ipynb: https://storage.googleapis.com/codyfalkosky/TensorFlowIssue/strange%20M1%20GPU%20Behavior.ipynb
```

**Comment by sushreebarsa** [comment]
@codyfalkosky Thank you for reporting the issue!
The  provided link for the code is not so could you please provide a gist or code snippet to replicate the issue?
Thank you!

**Comment by codyfalkosky** [comment]
#!/usr/bin/env python
# coding: utf-8

import numpy as np
import matplotlib.pyplot as plt

import tensorflow as tf

#########
# comment and uncomment the following line out after resetting the kernel to see different CPU/GPU behavior
# tf.config.set_visible_devices([], 'GPU')


def run_test():
    
    # Data
    # data for model training
    # a simple quadratic equation
    x = np.linspace(-10, 10, 10).reshape((-1, 1))
    y = 4*x**2 + 7
    
    # model
    model = tf.keras.models.Sequential([
        tf.keras.layers.Dense(16, activation='relu', input_shape=(1,)),
        tf.keras.layers.Dense(16, activation='relu'),
        tf.keras.layers.Dense(16, activation='relu'),
        tf.keras.layers.Dense(1)        
    ])
    
    # training objects
    opt = tf.keras.optimizers.legacy.Adam(0.001)
    los = tf.keras.losses.MeanSquaredError()
    met = tf.keras.metrics.MeanAbsoluteError()
    
    # compile
    model.compile(optimizer=opt, loss=los, metrics=[met])
    
    # train
    x_len = len(x)
    history = model.fit(x, y, epochs=5000, batch_size=x_len, verbose=0)
    
    # predict
    y_pred = model.predict(x, verbose=0)
    
    # plot
    plt.figure(figsize=(10,4))

    # Model Output Graph
    plt.subplot(1, 2, 1)
    plt.plot(x, y, label='Actual Data')
    plt.plot(x, y_pred, label='Model Output')
    plt.title('Model Output vs. Actual')
    plt.xlabel('x')
    plt.ylabel('y')
    plt.legend()

    # Loss History Graph
    plt.subplot(1, 2, 2)
    plt.plot(history.history['loss'])
    plt.xlabel('Epochs')
    plt.ylabel('MSE Loss')
    plt.title('Model MSE Loss History')
    
    plt.suptitle('Mac M1 CPU Behavior')
    for device in tf.config.get_visible_devices():
        if device.device_type == 'GPU':
            plt.suptitle('Mac M1 GPU Behavior')
            break

    plt.subplots_adjust(wspace=.3)
    plt.show()    


run_test()

**Comment by sushreebarsa** [comment]
@codyfalkosky I tried to replicate the issue so could you please have a look at the [gist](https://colab.research.google.com/gist/sushreebarsa/18ccf79b04db3d8da36ffdb919cd083f/61772.ipynb) and let me know if I am missing something?
Thank you!

**Comment by codyfalkosky** [comment]
link to corrected code:[https://colab.research.google.com/drive/1gWu5QNv5lLdPRMzry4ISTOh6_TU0uwFW](url)

This error is only when processing on a Mac M1 Max, so running in a colab will not duplicate the issue, as it is hardware related and must be run on an M1 Max MacBook Pro to duplicate the issue.

**Comment by codyfalkosky** [comment]
here is a repaired link to the original file if that is helpful: [https://storage.googleapis.com/codyfalkosky/TensorFlowIssue/strange%20M1%20GPU%20Behavior.ipynb](url)

**Comment by SuryanarayanaY** [comment]
@codyfalkosky ,

Could you please submit the exact link. Above links are not working. Thank you!

**Comment by codyfalkosky** [comment]
link:
https://storage.googleapis.com/codyfalkosky/TensorFlowIssue/strange%20M1%20GPU%20Behavior.ipynb

Thank you,
Cody

**Comment by SuryanarayanaY** [comment]
Hi @codyfalkosky ,

Apologies for the delay. I have tested the same code with TF2.15v and its working as intended.Please refer to attached snapshot below.

<img width="1031" alt="Screenshot 2023-11-28 at 16 21 44" src="https://github.com/tensorflow/tensorflow/assets/116063290/3ad1cc5c-b27d-4608-8056-4b94970462a6">

Could you please verify with TF2.15v and let us know if issue resolved for you also. Thanks!
